llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp

```diff
@@ -2176,6 +2176,27 @@ static VPRecipeBase *optimizeMaskToEVL(VPValue *HeaderMask,
       .Default([&](VPRecipeBase *R) { return nullptr; });
 }
 
+static void convertToEVLReverse(VPlan &Plan, VPTypeAnalysis &TypeInfo,
+                                VPValue &AllOneMask, VPValue &EVL) {
+  for (VPBasicBlock *VPBB : VPBlockUtils::blocksOnly<VPBasicBlock>(
+           vp_depth_first_shallow(Plan.getVectorLoopRegion()->getEntry()))) {
+    for (VPRecipeBase &R : make_early_inc_range(reverse(*VPBB))) {
+      auto *VPI = dyn_cast<VPInstruction>(&R);
+      if (!VPI || VPI->getOpcode() != VPInstruction::Reverse)
+        continue;
+
+      SmallVector<VPValue *> Ops(VPI->operands());
+      Ops.append({&AllOneMask, &EVL});
+      auto *NewReverse = new VPWidenIntrinsicRecipe(
+          Intrinsic::experimental_vp_reverse, Ops,
+          TypeInfo.inferScalarType(VPI), VPI->getDebugLoc());
```

Comment:
Minor nit, you could also use pattern matching here to avoid the dyn_cast? I'm not strongly opinionated on this though, happy to go with what you prefer

```suggestion
      VPValue *X;
      if (!match(&R, m_VPInstruction<VPInstruction::Reverse>(m_VPValue(X)))
        continue;
      auto *NewReverse = new VPWidenIntrinsicRecipe(
          Intrinsic::experimental_vp_reverse, {X, &AllOnesMask, &EVL},
          TypeInfo.inferScalarType(R.getVPSingleValue()), R.getDebugLoc());
```